kvdb/sqlbase/db.go

```diff
@@ -227,112 +215,43 @@ func (db *db) Update(f func(tx walletdb.ReadWriteTx) error,
 	return db.executeTransaction(f, reset, false)
 }
 
-// randRetryDelay returns a random retry delay between -50% and +50% of the
-// configured delay that is doubled for each attempt and capped at a max value.
-func randRetryDelay(initialRetryDelay, maxRetryDelay time.Duration,
-	attempt int) time.Duration {
-
-	halfDelay := initialRetryDelay / 2
-	randDelay := rand.Int63n(int64(initialRetryDelay)) //nolint:gosec
-
-	// 50% plus 0%-100% gives us the range of 50%-150%.
-	initialDelay := halfDelay + time.Duration(randDelay)
-
-	// If this is the first attempt, we just return the initial delay.
-	if attempt == 0 {
-		return initialDelay
-	}
-
-	// For each subsequent delay, we double the initial delay. This still
-	// gives us a somewhat random delay, but it still increases with each
-	// attempt. If we double something n times, that's the same as
-	// multiplying the value with 2^n. We limit the power to 32 to avoid
-	// overflows.
-	factor := time.Duration(math.Pow(2, math.Min(float64(attempt), 32)))
-	actualDelay := initialDelay * factor
-
-	// Cap the delay at the maximum configured value.
-	if actualDelay > maxRetryDelay {
-		return maxRetryDelay
-	}
-
-	return actualDelay
-}
-
 // executeTransaction creates a new read-only or read-write transaction and
 // executes the given function within it.
 func (db *db) executeTransaction(f func(tx walletdb.ReadWriteTx) error,
 	reset func(), readOnly bool) error {
 
-	// waitBeforeRetry is a helper function that will wait for a random
-	// interval before exiting to retry the db transaction. If false is
-	// returned, then this means that daemon is shutting down so we
-	// should abort the retries.
-	waitBeforeRetry := func(attemptNumber int) bool {
-		retryDelay := randRetryDelay(
-			DefaultInitialRetryDelay, DefaultMaxRetryDelay,
-			attemptNumber,
-		)
-
-		log.Debugf("Retrying transaction due to tx serialization "+
-			"error, attempt_number=%v, delay=%v", attemptNumber,
-			retryDelay)
-
-		select {
-		// Before we try again, we'll wait with a random backoff based
-		// on the retry delay.
-		case <-time.After(retryDelay):
-			return true
-
-		// If the daemon is shutting down, then we'll exit early.
-		case <-db.ctx.Done():
-			return false
-		}
+	makeTx := func() (sqldb.Tx, error) {
+		return newReadWriteTx(db, readOnly)
 	}
 
-	for i := 0; i < DefaultNumTxRetries; i++ {
-		reset()
-
-		tx, err := newReadWriteTx(db, readOnly)
-		if err != nil {
-			dbErr := MapSQLError(err)
-
-			if IsSerializationError(dbErr) {
-				// Nothing to roll back here, since we didn't
-				// even get a transaction yet.
-				if waitBeforeRetry(i) {
-					continue
-				}
-			}
-
-			return dbErr
+	execTxBody := func(tx sqldb.Tx) error {
+		kvTx, ok := tx.(*readWriteTx)
+		if !ok {
+			return fmt.Errorf("expected *readWriteTx, got %T", tx)
 		}
 
-		err = catchPanic(func() error { return f(tx) })
```

Comment:
I wonder if we want to port the `catchPanic` over to the common code as well? Since apparently the wallet sometimes panics? Or why was this added in the first place?

Reply:
Oops, never mind, I see that it's part of the `execTxBody` now...